<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Trang T. Le" />
  <meta name="author" content="Weixuan Fu" />
  <meta name="author" content="Jason H. Moore" />
  <meta name="dcterms.date" content="2018-12-19" />
  <meta name="keywords" content="tpot, automl, machine learning" />
  <title>Scaling tree-based automated machine learning to biomedical big data with a dataset selector</title>
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <link rel="stylesheet" href="github-pandoc.css" />
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!-- Insert Analytics Script Below -->
  <script>
  </script>
  <!-- End Analytics Script -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Scaling tree-based automated machine learning to biomedical big data with a dataset selector</h1>
</header>
<p><small><em> This manuscript (<a href="https://trang1618.github.io/tpot-ds-ms/v/85258915066ef1038c9812ee1c6b5942a154a5dc/">permalink</a>) was automatically generated from <a href="https://github.com/trang1618/tpot-ds-ms/tree/85258915066ef1038c9812ee1c6b5942a154a5dc">trang1618/tpot-ds-ms@8525891</a> on December 19, 2018. </em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Trang T. Le</strong><sup>☯</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0003-3737-6565">0000-0003-3737-6565</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/trang1618">trang1618</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/trang1618">trang1618</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 </small></p></li>
<li><p><strong>Weixuan Fu</strong><sup>☯</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0002-6434-5468">0000-0002-6434-5468</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/weixuanfu">weixuanfu</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/weixuanfu">weixuanfu</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 </small></p></li>
<li><p><strong>Jason H. Moore</strong><sup>†</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0002-5015-1099">0000-0002-5015-1099</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/EpistasisLab">EpistasisLab</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/moorejh">moorejh</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 · Funded by National Institute of Health Grant Nos. LM010098, LM012601 </small></p></li>
</ul>
<p><sup>☯</sup> — These authors contributed equally to this work.</p>
<p><sup>†</sup> — Direct correspondence to jhmoore@upenn.edu.</p>
<h2 id="abstract" class="page_break_before">Abstract</h2>
<p>Automated machine learning (AutoML) systems are helpful data science assistants designed to scan data for novel features, select appropriate supervised learning models and optimize their parameters. For this purpose, Tree-based Pipeline Optimization Tool (TPOT) was developed using strongly typed genetic programming to recommend an optimized analysis pipeline for the data scientist’s prediction problem. However, TPOT may reach computational resource limits when working on big data such as whole-genome expression data. We introduce two new features implemented in TPOT that helps increase the system’s scalability: Dataset selector and Template. Dataset selector (DS) provides the option to specify subsets of the features as separate datasets, assuming the signals come from one or more of these specific data subsets. Built in at the beginning of each pipeline structure, DS reduces the computational expense of TPOT to only evaluate on a smaller subset of data rather than the entire dataset. Consequently, DS increases TPOT’s efficiency in application on big data by slicing the dataset into smaller sets of features and allowing genetic programming to select the best subset in the final pipeline. Template enforces type constraints with strongly typed genetic programming and enables the incorporation of DS at the beginning of each pipeline. We show that DS and Template help reduce TPOT computation time and may provide more interpretable results. Our simulations show TPOT-DS significantly outperforms a tuned XGBoost model and standard TPOT implementation. We apply TPOT-DS to real RNA-Seq data from a study of major depressive disorder. Independent of the previous study that identified significant association with depression severity of the enrichment scores of two modules, in an automated fashion, TPOT-DS corroborates that one of the modules is largely predictive of the clinical diagnosis of each individual.</p>
<h2 id="author-summary">Author Summary</h2>
<p>Big data have recently become prevalent in many fields including meteorology, complex physics simulations, large scale imaging, genomics, biomedical research, environmental research and more. TPOT is a Python Automated Machine Learning (AutoML) tool that uses genetic programming to optimize machine learning pipelines for analyzing biomedical data. However, like other AutoML tools, when analyzing big data, the early implementations of TPOT face the challenges of long runtime, high computational expense as well complex pipeline with low interpretability. Here, we develop two novel features for TPOT, Dataset Selector and Template, that leverage domain knowledge, greatly reduce the computational expense and flexibly extend TPOT’s application to biomedical big data analysis.</p>
<h2 id="introduction">Introduction</h2>
<p>For many bioinformatics problems of classifying individuals into clinical categories from high-dimensional biological data, performance of a machine learning (ML) model depends greatly on the problem it is applied to <span class="citation" data-cites="BnXfrP0P h0b16kU4">[<a href="#ref-BnXfrP0P">1</a>,<a href="#ref-h0b16kU4">2</a>]</span>. In addition, choosing a classifier is merely one step of the arduous process that leads to predictions. To detect patterns among features (<em>e.g.</em>, clinical variables) and their associations with the outcome (<em>e.g.</em>, clinical diagnosis), a data scientist typically has to design and test different complex machine learning (ML) frameworks that consist of data exploration, feature engineering, model selection and prediction. Automated machine learning (AutoML) systems were developed to automate this challenging and time-consuming process. These intelligent systems increase the accessibility and scalability of various machine learning applications by efficiently solving an optimization problem to discover pipelines that yield satisfactory outcomes, such as prediction accuracy. Consequently, AutoML allows data scientists to focus their effort in applying their expertise in other important research components such as developing meaningful hypotheses or communicating the results.</p>
<p>Various approaches have been employed to build AutoML systems for diverse applications. Auto-sklearn <span class="citation" data-cites="8JQDv397">[<a href="#ref-8JQDv397">3</a>]</span> and Auto-WEKA <span class="citation" data-cites="S6aZVb3n">[<a href="#ref-S6aZVb3n">4</a>]</span> use Bayesian optimization for model selection and hyperparameter optimization. Recipe <span class="citation" data-cites="6ChydIkb">[<a href="#ref-6ChydIkb">6</a>]</span> optimizes the ML pipeline through grammar-based genetic programming and Autostacker <span class="citation" data-cites="RiocGZOq">[<a href="#ref-RiocGZOq">7</a>]</span> automates stacked ensembling. Both methods automate hyperparameter tuning and model selection using evolutionary algorithm. DEvol <span class="citation" data-cites="1CQSBxFFr">[<a href="#ref-1CQSBxFFr">8</a>]</span> designs deep neural network specifically via genetic programming. H2O.ai <span class="citation" data-cites="sOdEzGT3">[<a href="#ref-sOdEzGT3">9</a>]</span> automates data preprocessing, hyperparameter tuning, random grid search and stacked ensembles in a distributed ML platform in multiple languages. Finally, Xcessiv <span class="citation" data-cites="HA8l3lpi">[<a href="#ref-HA8l3lpi">10</a>]</span> provides web-based application for quick, scalable, and automated hyper-parameter tuning and stacked ensembling in Python.</p>
<p>Tree-based Pipeline Optimization Tool (TPOT) is a genetic programming-based AutoML system that uses genetic programming (GP) <span class="citation" data-cites="NopW1Vw3">[<a href="#ref-NopW1Vw3">11</a>]</span> to optimize a series of feature selectors, preprocessors and ML models with the objective of maximizing classification accuracy. While most AutoML systems primarily focus on model selection and hyperparameter optimization, TPOT also pays attention to feature selection and feature engineering by evaluating the complete pipelines based on their cross-validated score such as mean squared error or balanced accuracy. Given no a priori knowledge about the problem, TPOT has been shown to frequently outperform standard machine learning analyses <span class="citation" data-cites="QkGSlAB3 JEn7WIoN">[<a href="#ref-QkGSlAB3">12</a>,<a href="#ref-JEn7WIoN">13</a>]</span>. Effort has been made to specialize TPOT for human genetics research, resulting in a useful extended version of TPOT, TPOT-MDR, that features Multifactor Dimensionality Reduction and an Expert Knowledge Filter <span class="citation" data-cites="AvvI4W9K">[<a href="#ref-AvvI4W9K">14</a>]</span>. However, at the current stage, TPOT still requires great computational expense to analyze large datasets such as in genome-wide association studies (GWAS) or gene expression analyses. Consequently, the application of TPOT on real-world datasets has been limited to small sets of features <span class="citation" data-cites="3LGbkjqK">[<a href="#ref-3LGbkjqK">15</a>]</span>.</p>
<p>In this work, we introduce two new features implemented in TPOT that helps increase the system’s scalability. First, the Dataset Selector (DS) allows the users to pass specific subsets of the features, reducing the computational expense of TPOT at the beginning of each pipeline to only evaluate on a smaller subset of data rather than the entire dataset. Consequently, DS increases TPOT’s efficiency in application on large data sets by slicing the data into smaller sets of features (<em>e.g.</em> genes) and allowing a genetic algorithm to select the best subset in the final pipeline. Second, Template enables the option for strongly typed GP, a method to enforce type constraints in genetic programming. By letting users specify a desired structure of the resulting machine learning pipeline, Template helps reduce TPOT computation time and potentially provide more interpretable results.</p>
<h2 id="methods">Methods</h2>
<p>We begin with descriptions of the two novel additions to TPOT, Dataset Selector and Template. Then, we provide detail of a real-world RNA-Seq expression dataset and describe a simulation approach to generate data comparable to the expression data. Finally, we discuss other methods and performance metrics for comparison. Detailed simulation and analysis code needed to reproduce the results has been made available on the GitHub repository <a href="https://github.com/trang1618/tpot-ds">https://github.com/trang1618/tpot-ds</a>.</p>
<h3 id="tree-based-pipeline-optimization-tool">Tree-based Pipeline Optimization Tool</h3>
<p>Tree-based Pipeline Optimization Tool (TPOT) automates the laborious process of designing a ML pipeline by representing pipelines as binary expression trees with ML operators as primitives. Pipeline elements include algorithms from the extensive library of scikit-learn <span class="citation" data-cites="1iuWTU7i">[<a href="#ref-1iuWTU7i">16</a>]</span> as well as other efficient implementations such as extreme gradient boosting. Applying GP with the NSGA-II Pareto optimization <span class="citation" data-cites="iBP5Naag">[<a href="#ref-iBP5Naag">17</a>]</span>, TPOT optimizes the accuracy achieved by the pipeline while accounting for its complexity. Specifically, to automatically generate and optimize these machine learning pipelines, TPOT utilizes the Python package DEAP <span class="citation" data-cites="Gcs0HrMy">[<a href="#ref-Gcs0HrMy">18</a>]</span> to implement the GP algorithm. Implementation details can be found at TPOT’s active Github repository <a href="https://github.com/EpistasisLab/tpot">https://github.com/EpistasisLab/tpot</a>.</p>
<h3 id="dataset-selector">Dataset Selector</h3>
<p>TPOT’s current operators include sets of feature pre-processors, feature transformers, feature selection techniques, and supervised classifiers and regressions. In this study, we introduce a new operator called Dataset Selector (DS) that enables biologically guided group-level feature selection. Specifically, taking place at the very first stage of the pipeline, DS passes only a specific subset of the features onwards, effectively slicing the large original dataset into smaller ones. Hence, with DS, users can specify subsets of features of interest to reduce the feature space’s dimension at pipeline initialization. From predefined subsets of features, the DS operator allows TPOT to select the best subset that maximizes average accuracy in <em>k</em>-fold cross validation (5-fold by default).</p>
<p>For example, in a gene expression analysis of major depressive disorder, a neuroscientist can specify collections of genes in pathways of interest and identify the important collection that helps predict the depression severity. Similarly, in a genome-wide association study of breast cancer, an analyst may assign variants in the data to different subsets of potentially related variants and detect the subset associated with the breast cancer diagnosis. In general, the DS operator allows for compartmentalization of the feature space to smaller subsets based on <em>a priori</em> expert knowledge about the biomedical dataset. From here, TPOT learns and selects the most relevant group of features for outcome prediction.</p>
<h3 id="template">Template</h3>
<p>Parallel with the establishment of the Dataset Selector operator, we now offer TPOT users the option to define a Template that provides a way to specify a desired structure for the resulting machine learning pipeline, which will reduce TPOT computation time and potentially provide more interpretable results.</p>
<p>Current implementation of Template supports linear pipelines, or path graphs, which are trees with two nodes (operators) of vertex degree 1, and the other <span class="math inline">\(n-2\)</span> nodes of vertex degree 2. Further, Template takes advantage of the strongly typed genetic programming framework that enforces data-type constraints <span class="citation" data-cites="KgFuJ0Jv">[<a href="#ref-KgFuJ0Jv">19</a>]</span> and imposes type-based restrictions on which element (<em>i.e.</em>, operator) type can be chosen at each node. In strongly typed genetic programming, while the fitness function and parameters remain the same, the initialization procedure and genetic operators (<em>e.g.</em>, mutation, crossover) must respect the enhanced legality constraints <span class="citation" data-cites="KgFuJ0Jv">[<a href="#ref-KgFuJ0Jv">19</a>]</span>. With a Template defined, each node in the tree pipeline is assigned one of the five major operator types: dataset selector, feature selection, feature transform, classifier or regressor. Moreover, besides the major operator types, each node can also be assigned more specifically as a method of an operator, such as decision trees for classifier. An example Template is Dataset selector → Feature transform → Decision trees.</p>
<h3 id="datasets">Datasets</h3>
<p>We apply TPOT with the new DS operator on both simulated datasets and a real world RNA-Seq gene expression dataset. With both real-world and simulated data, we hope to acquire a comprehensive view of the strengths and limitations of TPOT in the next generation sequencing domain.</p>
<h4 id="simulation-methods">Simulation methods</h4>
<p>The simulated datasets were generated using the <code>R</code> package <code>privateEC</code>, which was designed to simulate realistic effects to be expected in gene expression or resting-state fMRI data. In the current study, to be consistent with the real expression dataset (described below), we simulate interaction effect data with <em>m</em> = 200 individuals (100 cases and 100 controls) and <em>p</em> = 5,000 real-valued features with 4% functional (true positive association with outcome) for each training and testing set. Full details of the simulation approach can be found in Refs. <span class="citation" data-cites="p7dAO241 NKnMeQUs">[<a href="#ref-p7dAO241">20</a>,<a href="#ref-NKnMeQUs">21</a>]</span>. Briefly, the privateEC simulation induces a differential co-expression network of random normal expression levels and permutes the values of targeted features within the cases to generate interactions. Further, by imposing a large number of background features (no association with outcome), we seek to assess TPOT-DS’s performance in accommodating large numbers of non-predictive features.</p>
<p>To closely resemble the module size distribution in the RNA-Seq data, we first fit a <span class="math inline">\(\Gamma\)</span> distribution to the observed module sizes then sample from this distribution values for the simulated subset size, before the total number of features reaches 4,800 (number of background features). Then, the background features were randomly placed in each subset corresponding to its size. Also, for each subset <span class="math inline">\(S_i, i = 1, \dots, n\)</span>, a functional feature <span class="math inline">\(s_j\)</span> belongs to the subset with the probability <span id="eq:p_subset" style="display: inline-block; position: relative; width: 100%"><span class="math display">\[P(s_j \in S_i) \sim 1.618^{-i}\]</span><span style="position: absolute; right: 0em; top: 50%; line-height:0; text-align: right">(1)</span></span>  where 1.618 is an approximation of the golden ratio and yields a reasonable distribution of the functional features: they are more likely to be included in the earlier subsets (subset 1 and 2) than the later ones.</p>
<h4 id="real-world-rna-seq-expression-data">Real-world RNA-Seq expression data</h4>
<p>We employed TPOT-DS on an RNA-Seq expression dataset of 78 individuals with major depressive disorder (MDD) and 79 healthy controls (HC) from Ref. <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span>. Gene expression levels were quantified from reads of 19,968 annotated protein-coding genes and underwent a series of preprocessing steps including low read-count and outlier removal, technical and batch effect adjustment, and coefficient of variation filtering. Consequently, whole blood RNA-Seq measurements of 5,912 genes were obtained and are now used in the current study to test for association with MDD status. We use the 23 subsets of interconnected genes called depression gene modules (DGMs) identified from the RNA-Seq gene network module analysis <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span> as input for the DS operator. We remark that these modules were constructed by an unsupervised machine learning method with dynamic tree cutting from a co-expression network. As a result, this prior knowledge of the gene structure does not depend on the diagnostic phenotype and thus yields no bias in the downstream analysis of TPOT-DS.</p>
<h3 id="performance-assessment">Performance assessment</h3>
<p>For each simulated and real-world dataset, after randomly splitting the entire data in two balanced smaller sets (75% training and 25% holdout), we trained TPOT-DS with the Template <code>Dataset Selector-Transformer-Classifier</code> on training data to predict class (<em>e.g.</em>, diagnostic phenotype in real-world data) in the holdout set. We assess the performance of TPOT-DS by quantifying its ability to correctly select the most important subset (containing most functional features) in 100 replicates of TPOT runs on simulated data with known underlying truth. To prevent potential overfitting, we select the pipeline that is closest to the 90th percentile of the cross-validation accuracy to be optimal. This rationale is motivated by a similar procedure for optimizing the penalty coefficient in regularized regression where the most parsimonious model within one standard error of the minimum cross-validation error is picked []. We compare the holdout (out-of-sample) accuracy of TPOT-DS’s optimal pipeline on the holdout set with that of standard TPOT (with <code>Transformer-Classifier</code> Template, no DS operator) and eXtreme Gradient Boosting <span class="citation" data-cites="8w9fI63O">[<a href="#ref-8w9fI63O">22</a>]</span>, or XGBoost, which is a fast and an efficient implementation of the gradient tree boosting method that has shown much utility in many winning Kaggle solutions <span class="citation" data-cites="1MHQyfXY">[<a href="#ref-1MHQyfXY">23</a>]</span> and been successfully incorporated in several neural network architectures <span class="citation" data-cites="19eUrsX1M 13as7dipI">[<a href="#ref-19eUrsX1M">24</a>,<a href="#ref-13as7dipI">25</a>]</span>. In the family of gradient boosted decision trees, XGBoost accounts for complex non-linear interaction structure among features and leverages gradient descents and boosting (sequential ensemble of weak classifiers) to effectively produce a strong prediction model. To obtain the optimal performance for this baseline model, we tune XGBoost hyperparameters using TPOT Template with only one classifier <code>XGBClassifier</code>, which is imported from the <code>xgboost</code> python package. Because of stochasticity in the optimal pipeline from TPOT-DS, standard TPOT and the tuned XGBoost model, we fit these models on the training data 100 times and compare 100 holdout accuracy values from each method. We choose accuracy to be the metric for comparison because phenotype is balanced in both simulated data and real-world data.</p>
<h3 id="manuscript-drafting">Manuscript drafting</h3>
<p>This manuscript is collaboratively written using Manubot <span class="citation" data-cites="1GGGHdsew">[<a href="#ref-1GGGHdsew">26</a>]</span>, a software that supports open paper writing via GitHub using the Markdown language. Manubot uses continuous integration to monitor changes and automatically update the manuscript. Consequently, the latest version of this manuscript is always available at <a href="https://trang1618.github.io/tpot-ds-ms/">https://trang1618.github.io/tpot-ds-ms/</a>.</p>
<h2 id="results">Results</h2>
<p>Our main goal is to test the performance of methods to identify features that discriminate between groups and optimize the classification accuracy.</p>
<h3 id="the-general-workflow-of-tpot-ds">The general workflow of TPOT-DS</h3>
<p>The template <code>Dataset Selector-Transformer-Classifier</code> specifies a simple general workflow of TPOT-DS (Fig. <a href="#fig:flow">1</a>).</p>
<figure>
<img src="images/flow.png" alt="Figure 1: TPOT-DS’s workflow and example pipelines. Final pipelines with optimized parameters are shown for simulated data (top) and real-world gene expression data (bottom)." id="fig:flow" style="width:90.0%" /><figcaption><span>Figure 1:</span> TPOT-DS’s workflow and example pipelines. Final pipelines with optimized parameters are shown for simulated data (top) and real-world gene expression data (bottom).</figcaption>
</figure>
<p>As discussed earlier in the Methods section, the optimal pipeline from TPOT-DS and standard TPOT is selected to be closest to the 90th percentile of the cross-validation accuracy. The optimal model of XGBoost holds properly tuned hyperparameters. For simulated dataset, the optimal pipeline selects subset <span class="math inline">\(S_1\)</span> then constructs an approximate feature map for a linear kernel with Nystroem, which uses a subset of the data as the basis for the approximation. The final prediction is made with an extra-trees classifier that fits a number of randomized decision trees on various sub-samples of the dataset with the presented optimized parameters (Fig. <a href="#fig:flow">1</a>). For the real-world dataset, the most optimal pipeline selects subset DGM-5, one-hot encode the features, then, similar to simulated data, makes the final prediction with an extra-trees classifier with a different set of optimized parameters (Fig. <a href="#fig:flow">1</a>).</p>
<h3 id="accuracy-assessment-of-optimal-pipelines">Accuracy assessment of optimal pipelines</h3>
<p>We compare the accuracy produced by optimal models from TPOT-DS, standard TPOT and XGBoost on classifying a simulated dataset with moderate interaction effect. We assign values of the effect size in the simulations to generate adequately challenging datasets so that the methods’ accuracies stay moderate and do not cluster around 0.5 or 1. The data set is split into 75% training and 25% holdout. The three models are built from the training dataset, then the trained model is applied to the independent holdout data to obtain the holdout accuracy.</p>
<p>We also apply the three methods to the RNA-Seq study of 78 major depressive disorder (MDD) subjects and 79 healthy controls (HC) described in <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span>. The dataset contains 5,912 genes after preprocessing and filtering (see Methods for more detail). We excluded 277 genes that did not belong to 23 subsets of interconnected genes (DGMs) so that the dataset remains the same across the three methods. As with simulated data, all models are built from the training dataset (61 HC, 56 MDD), then the trained model is applied to the independent holdout data (18 HC, 22 MDD).</p>
<p>For the simulated data, across all 100 model fits, the optimal TPOT-DS pipeline yields an average holdout prediction accuracy of 0.65, while the standard TPOT without DS and tuned XGBoost models respectively report an average holdout accuracy of 0.48 and 0.49 (Fig. <a href="#fig:compAcc">2</a>). This overfitting in the performance of these other two models is likely due to the models’ high flexibility that <em>over-learns</em> the training data, especially with the presence of many noisy background features.</p>
<figure>
<img src="images/compareAcc.svg" alt="Figure 2: Performance comparison of three models: tuned XGBoost, optimal pipeline from standard TPOT and optimal pipeline from TPOT-DS." id="fig:compAcc" style="width:90.0%" /><figcaption><span>Figure 2:</span> Performance comparison of three models: tuned XGBoost, optimal pipeline from standard TPOT and optimal pipeline from TPOT-DS.</figcaption>
</figure>
<p>Meanwhile, for the real-world expression data, the optimal TPOT-DS pipeline yields an average holdout prediction accuracy of 0.68, while the standard TPOT without DS and tuned XGBoost models produce average holdout accuracies of 0.60 and 0.59 respectively across all 100 model fits (Fig. <a href="#fig:compAcc">2</a>). In summary, the optimal models from standard TPOT and XGBoost perform better in real-world data compared to simulated data but still worse than that of TPOT-DS. In both datasets, separate Welch two-sample one-sided <em>t</em>-tests show TPOT-DS optimal pipelines significantly outperform those of XGBoost and standard TPOT (all <em>p</em> values <span class="math inline">\(&lt;10^{-15}\)</span>).</p>
<h3 id="consistency-in-selecting-subsets-of-tpot-ds">Consistency in selecting subsets of TPOT-DS</h3>
<figure>
<img src="images/sim_100.svg" alt="Figure 3: TPOT-DS’s holdout accuracy in simulated data with selected subset. Number of pipeline inclusions of each subset in 100 replications is displayed above the boxplots. Subset S_1 is the most frequent to be included in the final pipeline and yields the best prediction accuracy in the holdout set." id="fig:simDS" style="width:100.0%" /><figcaption><span>Figure 3:</span> TPOT-DS’s holdout accuracy in simulated data with selected subset. Number of pipeline inclusions of each subset in 100 replications is displayed above the boxplots. Subset <span class="math inline">\(S_1\)</span> is the most frequent to be included in the final pipeline and yields the best prediction accuracy in the holdout set.</figcaption>
</figure>
<p>Our simulation design produces a reasonable distribution of the functional features in all subsets, of which proportions are shown in Table [S1]. According to Eq. <a href="#eq:p_subset">1</a>, the earlier the subset, the more functional features it has. Therefore, our aim is to determine how well TPOT-DS can identify the first subset (<span class="math inline">\(S_1\)</span>) that contains the largest number of informative features. In 100 replications, TPOT-DS correctly selects subset <span class="math inline">\(S_1\)</span> in 75 resulting pipelines (Fig. <a href="#fig:simDS">3</a>), with the highest average holdout accuracy (0.69 across all 75 pipelines).</p>
<figure>
<img src="images/real_100.svg" alt="Figure 4: TPOT-DS’s holdout accuracy in RNA-Seq expression data with selected subset. Number of pipeline inclusions of each subset in 100 replications is displayed above the boxplots. Subsets DGM-5 and DGM-13 are the most frequent to be included in the final pipeline. Pipelines that include DGM-5, on average, produce higher MDD prediction accuracies in the holdout set." id="fig:realDS" style="width:80.0%" /><figcaption><span>Figure 4:</span> TPOT-DS’s holdout accuracy in RNA-Seq expression data with selected subset. Number of pipeline inclusions of each subset in 100 replications is displayed above the boxplots. Subsets DGM-5 and DGM-13 are the most frequent to be included in the final pipeline. Pipelines that include DGM-5, on average, produce higher MDD prediction accuracies in the holdout set.</figcaption>
</figure>
<p>For the expression data, in 100 replications, TPOT-DS selects DGM-5 (291 genes) 64 times to be the subset most predictive of the diagnosis status (Fig. <a href="#fig:realDS">4</a>), with the highest average holdout accuracy of 0.636 across 64 pipelines. In the previous study with a modular network approach, we showed that DGM-5 has statistically significant associations with depression severity measured by the Montgomery-Åsberg Depression Scale (MADRS). Although there is no direct link between the top genes of the module (Fig. <a href="#fig:featImp">5</a>a) and MDD in the literature, many of these genes interact with other MDD-related genes. For example, NR2C2 and TCF7L1 interact with FKBP5 gene whose association with MDD has been strongly suggested <span class="citation" data-cites="KXwvC8hd sxkRCGzQ rqdZ0HWl">[<a href="#ref-KXwvC8hd">27</a>,<a href="#ref-sxkRCGzQ">28</a>,<a href="#ref-rqdZ0HWl">29</a>]</span>. Many of DGM-5’s top genes were also shown to have statistically significant association with diagnosis phenotypes from a univariate analysis after multiple hypothesis testing correction <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span>. Further, with 82% overlap of DGM-5’s genes in a separate dataset from the RNA-Seq study by Mostafavi et al. <span class="citation" data-cites="g454CrrS">[<a href="#ref-g454CrrS">30</a>]</span>, this gene collection’s enrichment score was also shown to be significantly associated with the diagnosis status in this independent dataset.</p>
<figure>
<img src="images/importanceFeatures.svg" alt="Figure 5: Permutation importance scores of the top twenty expression features in the optimal pipeline that selects DGM-5 and one that selects DGM-13. Comprehensive importance scores of the all expression features computed by permutation from the optimal pipelines are provided in Table S2." id="fig:featImp" style="width:100.0%" /><figcaption><span>Figure 5:</span> Permutation importance scores of the top twenty expression features in the optimal pipeline that selects DGM-5 and one that selects DGM-13. Comprehensive importance scores of the all expression features computed by permutation from the optimal pipelines are provided in Table S2.</figcaption>
</figure>
<p>After DGM-5, DGM-13 (134 genes) was selected by TPOT-DS 30 times (Fig. <a href="#fig:realDS">4</a>), with an average holdout accuracy of 0.563 across 30 pipelines. The previous network approach did not find statistically significant association between this module’s enrichment score and the MADRS. Similar to DGM-5, in the literature, there is no direct link between MDD and the top genes in the module (Fig. <a href="#fig:featImp">5</a>b). Gene set enrichment analysis reported DGM-13’s involvement in axon guidance and developmental biology pathways with Reactome-FDR <em>q</em>-value <span class="math inline">\(&lt;\)</span> 0.05 <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span>.</p>
<h3 id="computational-expense">Computational expense</h3>
<p>For a dataset of the size simulated in our study (<em>m</em> = 200 samples and <em>p</em> = 5000 attributes), standard TPOT has a 18.5-hour runtime on a low performance computing machine with an Intel Xeon E5-2690 2.60GHz CPU, 28 cores and 256GB of RAM, whereas TPOT-DS has a 65-minute runtime, approximately 17 times faster. On the same low performance computing machine, each replication of standard TPOT on the expression data takes on average 13.3 hours, whereas TPOT-DS takes 40 minutes, approximately 20 times faster.</p>
<h2 id="discussion">Discussion</h2>
<p>To our knowledge, TPOT-DS is the first AutoML tool to offer the option of feature selection at the group level. Previously, it was computationally expensive for any AutoML program to process biomedical big data. TPOT-DS is able to identify the most meaningful group of features to include in the prediction pipeline. We assess TPOT-DS’s holdout prediction accuracy compared to standard TPOT and XGBoost, another state-of-the-art machine learning method. We apply TPOT-DS to real-world expression data to demonstrate the identification of biologically relevant groups of genes.</p>
<p>Implemented with a strongly typed GP, Template provides more flexibility by allowing users to pre-specify a particular pipeline structure based on their knowledge, which speeds up AutoML process and provides potentially more interpretable results. For example, in high-dimensional data, dimensionality reduction or feature selection algorithms are preferably included at the beginning of the pipelines via Template to identify important features and, meanwhile, reduce computation time. For datasets with categorical features, preprocessing operators for encoding those features such as one-hot encoder should be specified in the pipeline structure to improve pipelines’ performance. Template was utilized in this study to specify the DS as the first step of the pipeline, which enables the comparison between the two TPOT implementations, with and without DS.</p>
<p>We simulated data of the similar scale and challenging enough for the models to have similar predictive power as in the real-world RNA-Seq data. TPOT-DS correctly selects the subset with the most important features in the majority of replications and produces high average holdout accuracy of 0.69. In both simulated and RNASeq gene expression data, the final TPOT-DS pipeline outperforms that of standard TPOT and XGBoost. The low holdout accuracies of standard TPOT and XGBoost are expected because of the few signals in a high-dimenional feature space of the data. Meanwhile, TPOT-DS was able to find the more compact feature space to operate on, resulting in higher prediction accuracy and lower computational expense.</p>
<p>Interestingly enough, TPOT-DS repeatedly selects DGM-5 to include in the final pipeline. In a previous study, we showed DGM-5 and DGM-17 enrichment scores were significantly associated with depression severity <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span>. We also remarked that DGM-5 contains many genes that are biologically relevant or previously associated with mood disorders <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">20</a>]</span> and its enriched pathways such as apoptosis indicates a genetic signature of MDD pertaining to shrinkage of brain region-specific volume due to cell loss <span class="citation" data-cites="19yG9lS3X Okd6uiRx">[<a href="#ref-19yG9lS3X">31</a>,<a href="#ref-Okd6uiRx">32</a>]</span>.</p>
<p>TPOT-DS also selects DGM-13 as a potentially predictive group of features with smaller average holdout accuracy compared to DGM-5 (0.563 <span class="math inline">\(&lt;\)</span> 0.636). While many of the top genes do not have direct disease association in MalaCards, several have been linked to depression-like behavior in animal studies such as PPP1R16A <span class="citation" data-cites="XKNZFJkN">[<a href="#ref-XKNZFJkN">33</a>]</span>, CASKIN1 <span class="citation" data-cites="hU27Whmi">[<a href="#ref-hU27Whmi">34</a>]</span> and MXRA8 <span class="citation" data-cites="lAZSHU1O">[<a href="#ref-lAZSHU1O">35</a>]</span>. The RGL4 gene, a Ral guanine nucleotide dissociation stimulator, was found to have a rare protein disruptive variant in at least one suicide patient among 60 other mutations <span class="citation" data-cites="cuk31K3D">[<a href="#ref-cuk31K3D">36</a>]</span>. The lack of previously found association of these genes with the phenotype is likely because MDD is a complex disorder of heterogeneous etiology <span class="citation" data-cites="sd1g3RjP">[<a href="#ref-sd1g3RjP">37</a>]</span>. Hence, the clinical diagnosis is the accumulative result of coordinated variation of many genes in the module, especially ones with high importance scores. Future studies to refine and characterize genes in DGM-13 as well as DGM-5 may deploy expression quantitative trait loci (e-QTL) or interaction QTL analysis to discover disease-associated variants <span class="citation" data-cites="1T5OumxC">[<a href="#ref-1T5OumxC">38</a>]</span>.</p>
<p>Complexity-interpretability trade-off is an important topic to discuss in the context of AutoML. While arbitrarily-shaped pipelines may yield predictions competitive to human-level performance, these pipelines are often too complex to be interpretable. Vice versa, a simpler pipeline with defined steps of operators may be easier to interpret but yield suboptimal prediction accuracy. Finding the balance between pipeline complexity, model interpretation and generalization remains a challenging task for AutoML application in biomedical big data. With DS, in the terminology of evolutionary algorithm, each pipeline individual of a TPOT generation during optimization holds lower complexity due to the selected subset’s lower dimension compared to that of the entire dataset. We hope that, with the complexity reduction from imposing a strongly-type GP template and DS, a small loss in dataset-specific predictive accuracy can be compensated by considerable increase in interpretability and generalizability. In this study, the resulting TPOT-DS pipelines are more interpretable with only two simple optimized operators after the DS: a transformer and a classifier. In the case of the expression analysis, these pipelines also highlight two small sets of interconnected genes that contain candidates for MDD and related disorders. Additionally, complexity reduction results in more efficient computation, which is strongly desirable in biomedical big data analysis.</p>
<p>A limitation of the DS analysis is the required pre-definition of subsets prior to executing TPOT-DS. While this characteristic of an intelligent system is desirable when <em>a prior</em> knowledge on the biomedical data is available, it might pose as a challenge when this knowledge is inadequate, such as when analyzing data of a brand-new disease. Nevertheless, one can perform a clustering method such as <em>k</em>-means to group features prior to performing TPOT-DS on the data. Another limitation of the current implementation of TPOT-DS is its restricted ability to select only one subset. A future design to support tree structures for Template will enable TPOT-DS to identify more than one subset that have high predictive power of the outcome. A new operator that combines the data subsets will prove useful in this design. Extensions of TPOT-DS will also involve overlapping subsets, which will require pipeline complexity reformulation beyond the total number of operators included in a pipeline. Specifically, in the case of overlapping subsets, the number of features in the selected subset(s) is expected to be an element of the complexity calculation. Extension of TPOT-DS to GWAS is straightforward. However, because of the low predictive power of variants in current GWAS, alternative metrics beside accuracy, balanced accuracy or area under the receiving operator characteristic curve will need to be designed and included in the fitness function of TPOT’s evolutionary algorithm.</p>
<p>In this study, we developed two new operators for TPOT, Dataset Selector and Template, to enhance its performance on high-dimensional data by simplifying the pipeline structure and reducing the computational expense. Dataset Selector helps users leverage domain knowledge to narrow down important features for further interpretation, and Template largely increases flexibility of TPOT via customizing pipeline structure. Future extension and integration of these two operators have the potential to enrich the application of AutoML on different real world biomedical problems.</p>
<h2 id="references" class="page_break_before">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs">
<div id="ref-BnXfrP0P">
<p>1. <strong>PMLB: a large benchmark suite for machine learning evaluation and comparison</strong><br />
Randal S. Olson, William La Cava, Patryk Orzechowski, Ryan J. Urbanowicz, Jason H. Moore<br />
<em>BioData Mining</em> (2017-12) <a href="https://doi.org/gfrbw5">https://doi.org/gfrbw5</a><br />
DOI: <a href="https://doi.org/10.1186/s13040-017-0154-4">10.1186/s13040-017-0154-4</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/29238404">29238404</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC5725843">PMC5725843</a></p>
</div>
<div id="ref-h0b16kU4">
<p>2. <strong>Data-driven advice for applying machine learning to bioinformatics problems.</strong><br />
Randal S Olson, William La Cava, Zairah Mustahsan, Akshay Varik, Jason H Moore<br />
<em>Pacific Symposium on Biocomputing. Pacific Symposium on Biocomputing</em> (2018) <a href="https://www.ncbi.nlm.nih.gov/pubmed/29218881">https://www.ncbi.nlm.nih.gov/pubmed/29218881</a><br />
PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/29218881">29218881</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC5890912">PMC5890912</a></p>
</div>
<div id="ref-8JQDv397">
<p>3. <strong>Efficient and Robust Automated Machine Learning</strong><br />
Matthias Feurer, Aaron Klein, Katharina Eggensperger, Jost Springenberg, Manuel Blum, Frank Hutter<br />
<em>Advances in Neural Information Processing Systems 28</em> (2015) <a href="http://papers.nips.cc/paper/5872-efficient-and-robust-automated-machine-learning.pdf">http://papers.nips.cc/paper/5872-efficient-and-robust-automated-machine-learning.pdf</a></p>
</div>
<div id="ref-S6aZVb3n">
<p>4. <strong>Auto-WEKA: Combined Selection and Hyperparameter Optimization of Classification Algorithms</strong><br />
Chris Thornton, Frank Hutter, Holger H. Hoos, Kevin Leyton-Brown<br />
<em>arXiv</em> (2012-08-18) <a href="https://arxiv.org/abs/1208.3719v2">https://arxiv.org/abs/1208.3719v2</a></p>
</div>
<div id="ref-ai67wdhp">
<p>5. <strong>Auto-WEKA 2.0: Automatic model selection and hyperparameter optimization in WEKA</strong><br />
Kevin Leyton-Brown, Frank Hutter, Holger H. Hoos, Chris Thornton, Lars Kotthoff<br />
<em>Journal of Machine Learning Research</em> (2017) <a href="http://www.jmlr.org/papers/v18/16-261.html">http://www.jmlr.org/papers/v18/16-261.html</a></p>
</div>
<div id="ref-6ChydIkb">
<p>6. <strong>RECIPE: A Grammar-Based Framework for Automatically Evolving Classification Pipelines</strong><br />
Alex G. C. de Sá, Walter José G. S. Pinto, Luiz Otavio V. B. Oliveira, Gisele L. Pappa<br />
<em>Lecture Notes in Computer Science</em> (2017) <a href="https://doi.org/gfjzg2">https://doi.org/gfjzg2</a><br />
DOI: <a href="https://doi.org/10.1007/978-3-319-55696-3_16">10.1007/978-3-319-55696-3_16</a></p>
</div>
<div id="ref-RiocGZOq">
<p>7. <strong>Autostacker: A Compositional Evolutionary Learning System</strong><br />
Boyuan Chen, Harvey Wu, Warren Mo, Ishanu Chattopadhyay, Hod Lipson<br />
<em>arXiv</em> (2018-03-02) <a href="https://arxiv.org/abs/1803.00684v1">https://arxiv.org/abs/1803.00684v1</a></p>
</div>
<div id="ref-1CQSBxFFr">
<p>8. <strong>Genetic ConvNet architecture search with Keras. Contribute to joeddav/devol development by creating an account on GitHub</strong><br />
Joe Davison<br />
(2018-12-11) <a href="https://github.com/joeddav/devol">https://github.com/joeddav/devol</a></p>
</div>
<div id="ref-sOdEzGT3">
<p>9. <strong>AutoML: Automatic Machine Learning — H2O 3.22.0.2 documentation</strong><a href="http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html">http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html</a></p>
</div>
<div id="ref-HA8l3lpi">
<p>10. <strong>A web-based application for quick, scalable, and automated hyperparameter tuning and stacked ensembling in Python.: reiinakano/xcessiv</strong><br />
Reiichiro Nakano<br />
(2018-12-18) <a href="https://github.com/reiinakano/xcessiv">https://github.com/reiinakano/xcessiv</a></p>
</div>
<div id="ref-NopW1Vw3">
<p>11. <strong>Genetic Programming: An Introduction: on the Automatic Evolution of Computer Programs and Its Applications</strong><br />
Wolfgang Banzhaf, Frank D. Francone, Robert E. Keller, Nordin Peter<br />
<em>ACM Digital Library</em> (1998) <a href="https://dl.acm.org/citation.cfm?id=280485">https://dl.acm.org/citation.cfm?id=280485</a></p>
</div>
<div id="ref-QkGSlAB3">
<p>12. <strong>Evaluation of a Tree-based Pipeline Optimization Tool for Automating Data Science</strong><br />
Randal S. Olson, Nathan Bartley, Ryan J. Urbanowicz, Jason H. Moore<br />
<em>Proceedings of the 2016 on Genetic and Evolutionary Computation Conference - GECCO ’16</em> (2016) <a href="https://doi.org/gfgqv2">https://doi.org/gfgqv2</a><br />
DOI: <a href="https://doi.org/10.1145/2908812.2908918">10.1145/2908812.2908918</a></p>
</div>
<div id="ref-JEn7WIoN">
<p>13. <strong>Identifying and Harnessing the Building Blocks of Machine Learning Pipelines for Sensible Initialization of a Data Science Automation Tool</strong><br />
Randal S. Olson, Jason H. Moore<br />
<em>arXiv</em> (2016-07-29) <a href="https://arxiv.org/abs/1607.08878v1">https://arxiv.org/abs/1607.08878v1</a></p>
</div>
<div id="ref-AvvI4W9K">
<p>14. <strong>Toward the automated analysis of complex diseases in genome-wide association studies using genetic programming</strong><br />
Andrew Sohn, Randal S. Olson, Jason H. Moore<br />
<em>Proceedings of the Genetic and Evolutionary Computation Conference on - GECCO ’17</em> (2017) <a href="https://doi.org/gfgqv3">https://doi.org/gfgqv3</a><br />
DOI: <a href="https://doi.org/10.1145/3071178.3071212">10.1145/3071178.3071212</a></p>
</div>
<div id="ref-3LGbkjqK">
<p>15. <strong>Integrated machine learning pipeline for aberrant biomarker enrichment (i-mAB): characterizing clusters of differentiation within a compendium of systemic lupus erythematosus patients</strong><br />
Trang T. Le, Nigel O. Blackwood, Jaclyn N. Taroni, Weixuan Fu, Matthew K. Breitenstein<br />
<em>arXiv</em> (2018-03-08) <a href="https://arxiv.org/abs/1803.04487v1">https://arxiv.org/abs/1803.04487v1</a></p>
</div>
<div id="ref-1iuWTU7i">
<p>16. <strong>Scikit-learn: Machine Learning in Python</strong><br />
Fabian Pedregosa, Gaël Varoquaux, Alexandre Gramfort, Vincent Michel, Bertrand Thirion, Olivier Grisel, Mathieu Blondel, Peter Prettenhofer, Ron Weiss, Vincent Dubourg, … Édouard Duchesnay<br />
<em>Journal of Machine Learning Research</em> (2011-10) <a href="https://hal.inria.fr/hal-00650905">https://hal.inria.fr/hal-00650905</a></p>
</div>
<div id="ref-iBP5Naag">
<p>17. <strong>A fast and elitist multiobjective genetic algorithm: NSGA-II</strong><br />
K. Deb, A. Pratap, S. Agarwal, T. Meyarivan<br />
<em>IEEE Transactions on Evolutionary Computation</em> (2002-04) <a href="https://doi.org/bnw2vv">https://doi.org/bnw2vv</a><br />
DOI: <a href="https://doi.org/10.1109/4235.996017">10.1109/4235.996017</a></p>
</div>
<div id="ref-Gcs0HrMy">
<p>18. <strong>DEAP: Evolutionary Algorithms Made Easy</strong><br />
Christian Gagné, Marc Parizeau, Marc-André Gardner, François-Michel De Rainville, Félix-Antoine Fortin<br />
<em>Journal of Machine Learning Research</em> (2012) <a href="http://www.jmlr.org/papers/v13/fortin12a.html">http://www.jmlr.org/papers/v13/fortin12a.html</a></p>
</div>
<div id="ref-KgFuJ0Jv">
<p>19. <strong>Strongly Typed Genetic Programming</strong><br />
David J. Montana<br />
<em>Evolutionary Computation</em> (1995-06) <a href="https://doi.org/ct3mnb">https://doi.org/ct3mnb</a><br />
DOI: <a href="https://doi.org/10.1162/evco.1995.3.2.199">10.1162/evco.1995.3.2.199</a></p>
</div>
<div id="ref-p7dAO241">
<p>20. <strong>Identification and replication of RNA-Seq gene network modules associated with depression severity</strong><br />
Trang T. Le, Jonathan Savitz, Hideo Suzuki, Masaya Misaki, T. Kent Teague, Bill C. White, Julie H. Marino, Graham Wiley, Patrick M. Gaffney, Wayne C. Drevets, … Jerzy Bodurka<br />
<em>Translational Psychiatry</em> (2018-09-05) <a href="https://doi.org/gd7jx7">https://doi.org/gd7jx7</a><br />
DOI: <a href="https://doi.org/10.1038/s41398-018-0234-3">10.1038/s41398-018-0234-3</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/30185774">30185774</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC6125582">PMC6125582</a></p>
</div>
<div id="ref-NKnMeQUs">
<p>21. <strong>Differential co-expression network centrality and machine learning feature selection for identifying susceptibility hubs in networks with scale-free structure</strong><br />
Caleb A Lareau, Bill C White, Ann L Oberg, Brett A McKinney<br />
<em>BioData Mining</em> (2015-02-03) <a href="https://doi.org/gb5fpr">https://doi.org/gb5fpr</a><br />
DOI: <a href="https://doi.org/10.1186/s13040-015-0040-x">10.1186/s13040-015-0040-x</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/25685197">25685197</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4326454">PMC4326454</a></p>
</div>
<div id="ref-8w9fI63O">
<p>22. <strong>XGBoost</strong><br />
Tianqi Chen, Carlos Guestrin<br />
<em>Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining - KDD ’16</em> (2016) <a href="https://doi.org/gdp84q">https://doi.org/gdp84q</a><br />
DOI: <a href="https://doi.org/10.1145/2939672.2939785">10.1145/2939672.2939785</a></p>
</div>
<div id="ref-1MHQyfXY">
<p>23. <strong>Kaggle: Your Home for Data Science</strong><a href="https://www.kaggle.com/">https://www.kaggle.com/</a></p>
</div>
<div id="ref-19eUrsX1M">
<p>24. <strong>Short-Term Load Forecasting Using EMD-LSTM Neural Networks with a Xgboost Algorithm for Feature Importance Evaluation</strong><br />
Huiting Zheng, Jiabin Yuan, Long Chen<br />
<em>Energies</em> (2017-08-08) <a href="https://doi.org/gbtqbr">https://doi.org/gbtqbr</a><br />
DOI: <a href="https://doi.org/10.3390/en10081168">10.3390/en10081168</a></p>
</div>
<div id="ref-13as7dipI">
<p>25. <strong>A Novel Image Classification Method with CNN-XGBoost Model</strong><br />
Xudie Ren, Haonan Guo, Shenghong Li, Shilin Wang, Jianhua Li<br />
<em>Digital Forensics and Watermarking</em> (2017) <a href="https://doi.org/gfgvf3">https://doi.org/gfgvf3</a><br />
DOI: <a href="https://doi.org/10.1007/978-3-319-64185-0_28">10.1007/978-3-319-64185-0_28</a></p>
</div>
<div id="ref-1GGGHdsew">
<p>26. <strong>Open collaborative writing with Manubot</strong><br />
Daniel S. Himmelstein, David R. Slochower, Venkat S. Malladi, Casey S. Greene, Anthony Gitter<br />
(2018-12-18) <a href="https://greenelab.github.io/meta-review/">https://greenelab.github.io/meta-review/</a></p>
</div>
<div id="ref-KXwvC8hd">
<p>27. <strong>Variations in FKBP5 and BDNF genes are suggestively associated with depression in a Swedish population-based cohort</strong><br />
Catharina Lavebratt, Elin Åberg, Louise K. Sjöholm, Yvonne Forsell<br />
<em>Journal of Affective Disorders</em> (2010-09) <a href="https://doi.org/b726hn">https://doi.org/b726hn</a><br />
DOI: <a href="https://doi.org/10.1016/j.jad.2010.02.113">10.1016/j.jad.2010.02.113</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/20226536">20226536</a></p>
</div>
<div id="ref-sxkRCGzQ">
<p>28. <strong>Modulation of glucocorticoid receptor nuclear translocation in neurons by immunophilins FKBP51 and FKBP52: Implications for major depressive disorder</strong><br />
Erick T. Tatro, Ian P. Everall, Marcus Kaul, Cristian L. Achim<br />
<em>Brain Research</em> (2009-08) <a href="https://doi.org/b9cn9m">https://doi.org/b9cn9m</a><br />
DOI: <a href="https://doi.org/10.1016/j.brainres.2009.06.036">10.1016/j.brainres.2009.06.036</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/19545546">19545546</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2724600">PMC2724600</a></p>
</div>
<div id="ref-rqdZ0HWl">
<p>29. <strong>Polymorphisms in FKBP5 are associated with increased recurrence of depressive episodes and rapid response to antidepressant treatment</strong><br />
Elisabeth B Binder, Daria Salyakina, Peter Lichtner, Gabriele M Wochnik, Marcus Ising, Benno Pütz, Sergi Papiol, Shaun Seaman, Susanne Lucae, Martin A Kohli, … Bertram Muller-Myhsok<br />
<em>Nature Genetics</em> (2004-11-21) <a href="https://doi.org/bh28xx">https://doi.org/bh28xx</a><br />
DOI: <a href="https://doi.org/10.1038/ng1479">10.1038/ng1479</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/15565110">15565110</a></p>
</div>
<div id="ref-g454CrrS">
<p>30. <strong>Type I interferon signaling genes in recurrent major depression: increased expression detected by whole-blood RNA sequencing</strong><br />
S Mostafavi, A Battle, X Zhu, JB Potash, MM Weissman, J Shi, K Beckman, C Haudenschild, C McCormick, R Mei, … DF Levinson<br />
<em>Molecular Psychiatry</em> (2013-12-03) <a href="https://doi.org/f6qdpr">https://doi.org/f6qdpr</a><br />
DOI: <a href="https://doi.org/10.1038/mp.2013.161">10.1038/mp.2013.161</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/24296977">24296977</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC5404932">PMC5404932</a></p>
</div>
<div id="ref-19yG9lS3X">
<p>31. <strong>A meta-analysis examining clinical predictors of hippocampal volume in patients with major depressive disorder.</strong><br />
Margaret C McKinnon, Kaan Yucel, Anthony Nazarov, Glenda M MacQueen<br />
<em>Journal of psychiatry &amp; neuroscience : JPN</em> (2009-01) <a href="https://www.ncbi.nlm.nih.gov/pubmed/19125212">https://www.ncbi.nlm.nih.gov/pubmed/19125212</a><br />
PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/19125212">19125212</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2612082">PMC2612082</a></p>
</div>
<div id="ref-Okd6uiRx">
<p>32. <strong>Increased apoptosis in patients with major depression: A preliminary study.</strong><br />
E Eilat, S Mendlovic, A Doron, V Zakuth, Z Spirer<br />
<em>Journal of immunology (Baltimore, Md. : 1950)</em> (1999-07-01) <a href="https://www.ncbi.nlm.nih.gov/pubmed/10384158">https://www.ncbi.nlm.nih.gov/pubmed/10384158</a><br />
PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/10384158">10384158</a></p>
</div>
<div id="ref-XKNZFJkN">
<p>33. <strong>A Molecular Signature of Depression in the Amygdala</strong><br />
Etienne Sibille Ph.D., Yingjie Wang M.S., Jennifer Joeyen-Waldorf B.S., Chris Gaiteri B.S., Alexandre Surget Ph.D., Sunghee Oh B.S., Catherine Belzung Ph.D., George C. Tseng Ph.D., David A. Lewis M.D.<br />
<em>American Journal of Psychiatry</em> (2009-09) <a href="https://doi.org/fmbrrk">https://doi.org/fmbrrk</a><br />
DOI: <a href="https://doi.org/10.1176/appi.ajp.2009.08121760">10.1176/appi.ajp.2009.08121760</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/19605536">19605536</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2882057">PMC2882057</a></p>
</div>
<div id="ref-hU27Whmi">
<p>34. <strong>Distribution of Caskin1 protein and phenotypic characterization of its knockout mice using a comprehensive behavioral test battery</strong><br />
Tayo Katano, Keizo Takao, Manabu Abe, Maya Yamazaki, Masahiko Watanabe, Tsuyoshi Miyakawa, Kenji Sakimura, Seiji Ito<br />
<em>Molecular Brain</em> (2018-10-25) <a href="https://doi.org/gfrhrr">https://doi.org/gfrhrr</a><br />
DOI: <a href="https://doi.org/10.1186/s13041-018-0407-2">10.1186/s13041-018-0407-2</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/30359304">30359304</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC6202847">PMC6202847</a></p>
</div>
<div id="ref-lAZSHU1O">
<p>35. <strong>Effect of Acute Stressor and Serotonin Transporter Genotype on Amygdala First Wave Transcriptome in Mice</strong><br />
Christa Hohoff, Ali Gorji, Sylvia Kaiser, Edith Willscher, Eberhard Korsching, Oliver Ambrée, Volker Arolt, Klaus-Peter Lesch, Norbert Sachser, Jürgen Deckert, Lars Lewejohann<br />
<em>PLoS ONE</em> (2013-03-11) <a href="https://doi.org/gfkt9h">https://doi.org/gfkt9h</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0058880">10.1371/journal.pone.0058880</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/23536833">23536833</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3594195">PMC3594195</a></p>
</div>
<div id="ref-cuk31K3D">
<p>36. <strong>High-Coverage Whole-Exome Sequencing Identifies Candidate Genes for Suicide in Victims with Major Depressive Disorder</strong><br />
Dóra Tombácz, Zoltán Maróti, Tibor Kalmár, Zsolt Csabai, Zsolt Balázs, Shinichi Takahashi, Miklós Palkovits, Michael Snyder, Zsolt Boldogkői<br />
<em>Scientific Reports</em> (2017-08-02) <a href="https://doi.org/gbrpc3">https://doi.org/gbrpc3</a><br />
DOI: <a href="https://doi.org/10.1038/s41598-017-06522-3">10.1038/s41598-017-06522-3</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/28769055">28769055</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC5541090">PMC5541090</a></p>
</div>
<div id="ref-sd1g3RjP">
<p>37. <strong>Genetic Studies of Major Depressive Disorder: Why Are There No Genome-wide Association Study Findings and What Can We Do About It?</strong><br />
Douglas F. Levinson, Sara Mostafavi, Yuri Milaneschi, Margarita Rivera, Stephan Ripke, Naomi R. Wray, Patrick F. Sullivan<br />
<em>Biological Psychiatry</em> (2014-10) <a href="https://doi.org/gd8sv4">https://doi.org/gd8sv4</a><br />
DOI: <a href="https://doi.org/10.1016/j.biopsych.2014.07.029">10.1016/j.biopsych.2014.07.029</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/25201436">25201436</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4740915">PMC4740915</a></p>
</div>
<div id="ref-1T5OumxC">
<p>38. <strong>An interaction quantitative trait loci tool implicates epistatic functional variants in an apoptosis pathway in smallpox vaccine eQTL data</strong><br />
CA Lareau, BC White, AL Oberg, RB Kennedy, GA Poland, BA McKinney<br />
<em>Genes &amp; Immunity</em> (2016-04-07) <a href="https://doi.org/f9mvjk">https://doi.org/f9mvjk</a><br />
DOI: <a href="https://doi.org/10.1038/gene.2016.15">10.1038/gene.2016.15</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/27052692">27052692</a></p>
</div>
</div>
<script>
// AnchorJS minified version below.
// Source https://github.com/bryanbraun/anchorjs/blob/064abdd0987f305933ec4982af6d0c1cf2fd0814/anchor.js

/**
 * AnchorJS - v4.0.0 - 2017-06-02
 * https://github.com/bryanbraun/anchorjs
 * Copyright (c) 2017 Bryan Braun; Licensed MIT
 */
!function(A,e){"use strict";"function"==typeof define&&define.amd?define([],e):"object"==typeof module&&module.exports?module.exports=e():(A.AnchorJS=e(),A.anchors=new A.AnchorJS)}(this,function(){"use strict";function A(A){function e(A){A.icon=A.hasOwnProperty("icon")?A.icon:"",A.visible=A.hasOwnProperty("visible")?A.visible:"hover",A.placement=A.hasOwnProperty("placement")?A.placement:"right",A.class=A.hasOwnProperty("class")?A.class:"",A.truncate=A.hasOwnProperty("truncate")?Math.floor(A.truncate):64}function t(A){var e;if("string"==typeof A||A instanceof String)e=[].slice.call(document.querySelectorAll(A));else{if(!(Array.isArray(A)||A instanceof NodeList))throw new Error("The selector provided to AnchorJS was invalid.");e=[].slice.call(A)}return e}function n(){if(null===document.head.querySelector("style.anchorjs")){var A,e=document.createElement("style");e.className="anchorjs",e.appendChild(document.createTextNode("")),void 0===(A=document.head.querySelector('[rel="stylesheet"], style'))?document.head.appendChild(e):document.head.insertBefore(e,A),e.sheet.insertRule(" .anchorjs-link {   opacity: 0;   text-decoration: none;   -webkit-font-smoothing: antialiased;   -moz-osx-font-smoothing: grayscale; }",e.sheet.cssRules.length),e.sheet.insertRule(" *:hover > .anchorjs-link, .anchorjs-link:focus  {   opacity: 1; }",e.sheet.cssRules.length),e.sheet.insertRule(" [data-anchorjs-icon]::after {   content: attr(data-anchorjs-icon); }",e.sheet.cssRules.length),e.sheet.insertRule(' @font-face {   font-family: "anchorjs-icons";   src: url(data:n/a;base64,AAEAAAALAIAAAwAwT1MvMg8yG2cAAAE4AAAAYGNtYXDp3gC3AAABpAAAAExnYXNwAAAAEAAAA9wAAAAIZ2x5ZlQCcfwAAAH4AAABCGhlYWQHFvHyAAAAvAAAADZoaGVhBnACFwAAAPQAAAAkaG10eASAADEAAAGYAAAADGxvY2EACACEAAAB8AAAAAhtYXhwAAYAVwAAARgAAAAgbmFtZQGOH9cAAAMAAAAAunBvc3QAAwAAAAADvAAAACAAAQAAAAEAAHzE2p9fDzz1AAkEAAAAAADRecUWAAAAANQA6R8AAAAAAoACwAAAAAgAAgAAAAAAAAABAAADwP/AAAACgAAA/9MCrQABAAAAAAAAAAAAAAAAAAAAAwABAAAAAwBVAAIAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAMCQAGQAAUAAAKZAswAAACPApkCzAAAAesAMwEJAAAAAAAAAAAAAAAAAAAAARAAAAAAAAAAAAAAAAAAAAAAQAAg//0DwP/AAEADwABAAAAAAQAAAAAAAAAAAAAAIAAAAAAAAAIAAAACgAAxAAAAAwAAAAMAAAAcAAEAAwAAABwAAwABAAAAHAAEADAAAAAIAAgAAgAAACDpy//9//8AAAAg6cv//f///+EWNwADAAEAAAAAAAAAAAAAAAAACACEAAEAAAAAAAAAAAAAAAAxAAACAAQARAKAAsAAKwBUAAABIiYnJjQ3NzY2MzIWFxYUBwcGIicmNDc3NjQnJiYjIgYHBwYUFxYUBwYGIwciJicmNDc3NjIXFhQHBwYUFxYWMzI2Nzc2NCcmNDc2MhcWFAcHBgYjARQGDAUtLXoWOR8fORYtLTgKGwoKCjgaGg0gEhIgDXoaGgkJBQwHdR85Fi0tOAobCgoKOBoaDSASEiANehoaCQkKGwotLXoWOR8BMwUFLYEuehYXFxYugC44CQkKGwo4GkoaDQ0NDXoaShoKGwoFBe8XFi6ALjgJCQobCjgaShoNDQ0NehpKGgobCgoKLYEuehYXAAAADACWAAEAAAAAAAEACAAAAAEAAAAAAAIAAwAIAAEAAAAAAAMACAAAAAEAAAAAAAQACAAAAAEAAAAAAAUAAQALAAEAAAAAAAYACAAAAAMAAQQJAAEAEAAMAAMAAQQJAAIABgAcAAMAAQQJAAMAEAAMAAMAAQQJAAQAEAAMAAMAAQQJAAUAAgAiAAMAAQQJAAYAEAAMYW5jaG9yanM0MDBAAGEAbgBjAGgAbwByAGoAcwA0ADAAMABAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAH//wAP) format("truetype"); }',e.sheet.cssRules.length)}}this.options=A||{},this.elements=[],e(this.options),this.isTouchDevice=function(){return!!("ontouchstart"in window||window.DocumentTouch&&document instanceof DocumentTouch)},this.add=function(A){var i,o,s,c,r,a,h,l,u,d,f,g,p=[];if(e(this.options),"touch"===(g=this.options.visible)&&(g=this.isTouchDevice()?"always":"hover"),A||(A="h2, h3, h4, h5, h6"),0===(i=t(A)).length)return this;for(n(),o=document.querySelectorAll("[id]"),s=[].map.call(o,function(A){return A.id}),r=0;r<i.length;r++)if(this.hasAnchorJSLink(i[r]))p.push(r);else{if(i[r].hasAttribute("id"))c=i[r].getAttribute("id");else if(i[r].hasAttribute("data-anchor-id"))c=i[r].getAttribute("data-anchor-id");else{u=l=this.urlify(i[r].textContent),h=0;do{void 0!==a&&(u=l+"-"+h),a=s.indexOf(u),h+=1}while(-1!==a);a=void 0,s.push(u),i[r].setAttribute("id",u),c=u}d=c.replace(/-/g," "),(f=document.createElement("a")).className="anchorjs-link "+this.options.class,f.href="#"+c,f.setAttribute("aria-label","Anchor link for: "+d),f.setAttribute("data-anchorjs-icon",this.options.icon),"always"===g&&(f.style.opacity="1"),""===this.options.icon&&(f.style.font="1em/1 anchorjs-icons","left"===this.options.placement&&(f.style.lineHeight="inherit")),"left"===this.options.placement?(f.style.position="absolute",f.style.marginLeft="-1em",f.style.paddingRight="0.5em",i[r].insertBefore(f,i[r].firstChild)):(f.style.paddingLeft="0.375em",i[r].appendChild(f))}for(r=0;r<p.length;r++)i.splice(p[r]-r,1);return this.elements=this.elements.concat(i),this},this.remove=function(A){for(var e,n,i=t(A),o=0;o<i.length;o++)(n=i[o].querySelector(".anchorjs-link"))&&(-1!==(e=this.elements.indexOf(i[o]))&&this.elements.splice(e,1),i[o].removeChild(n));return this},this.removeAll=function(){this.remove(this.elements)},this.urlify=function(A){var t=/[& +$,:;=?@"#{}|^~[`%!'<>\]\.\/\(\)\*\\]/g;return this.options.truncate||e(this.options),A.trim().replace(/\'/gi,"").replace(t,"-").replace(/-{2,}/g,"-").substring(0,this.options.truncate).replace(/^-+|-+$/gm,"").toLowerCase()},this.hasAnchorJSLink=function(A){var e=A.firstChild&&(" "+A.firstChild.className+" ").indexOf(" anchorjs-link ")>-1,t=A.lastChild&&(" "+A.lastChild.className+" ").indexOf(" anchorjs-link ")>-1;return e||t||!1}}return A});

// Enable links for selected headers
var anchors = new AnchorJS();
anchors.add("h2, h3, h4")
</script>
<!-- Enable Hypothesis annotation. https://web.hypothes.is/ -->
<script src="https://hypothes.is/embed.js" async></script>
</body>
</html>
