## Discussion
To our knowledge, TPOT-DS is the first AutoML tool to offer the option of feature selection at the group level.
Previously, it was computationally expensive for any AutoML program to process biomedical big data.
TPOT-DS is able to identify the most meaningful group of features to include in the prediction pipeline. 
We assessed TPOT-DS's out-of-sample prediction accuracy compared to standard TPOT and XGBoost, another state-of-the-art machine learning method.
We applied TPOT-DS to real-world expression data to demonstrate the identification of biologically relevant groups of genes.

Implemented with a strongly typed GP, Template allows users to pre-specify a particular pipeline structure, which speeds up the automation computation time and provides potentially more interpretable results.
Consequently, Template enables the comparison between the two TPOT implementations, with and without DS.

We simulated data of the similar scale and chalenging enough for the models to have similar predictive power as in the real-world RNA-Seq data.
TPOT-DS correctly selects the first subset (containing the most important features) 75% of the time with high holdout accuracy (0.69).
When another subset is chosen in the final pipeline, this method still produces holdout accuracy comparable to that of standard TPOT and XGBoost (0.565 - 0.575).
For the RNASeq gene expression data, the best TPOT-DS pipeline selects DGM-5 and reports competitive accuracy with standard TPOT (0.636 vs. 0.642) but with a smaller feature space (291 vs. 5,635 genes) and 20 times more computationally efficient.
Both TPOT-DS and standard TPOT significantly outperform XGBoost in terms of accuracy (*p* = []).

Interestingly enough, TPOT-DS repeatedly selects DGM-5 to include in the final pipeline. 
In a previous study, we showed DGM-5 and DGM-17 enrichment scores were significantly associated with depression severity [@doi:10.1038/s41398-018-0234-3].
We also remarked that DGM-5 contains many genes that are biologically relevant or previously associated with mood disorders [@doi:10.1038/s41398-018-0234-3] and its enriched pathways such as apoptosis indicates a genetic signature of MDD pertaining shrinkage of brain region-specific volume due to cell loss [@pmid:19125212;@pmid:10384158].

TPOT-DS also selects DGM-13 as a potentially predictive group of features with smaller average holdout accuracy compared to DGM-5 (0.563 $<$ 0.636).
While many of the top genes do not have direct disease association in MalaCards, several have been linked to depression in animal studies such as PPP1R16A [@doi:10.1176/appi.ajp.2009.08121760] and MXRA8 [@doi:10.1371/journal.pone.0058880].
Further, the RGL4 gene, a Ral guanine nucleotide dissociation stimulator, was found to have a rare protein disruptive variant in at least one suicide patient among 60 other mutations [@doi:10.1038/s41598-017-06522-3].
[More here...]

Complexity-interpretability trade-off is an important topic to discuss in the context of AutoML.
While arbitrarily-shaped pipelines may yield predictions competitive to human-level performance, these pipelines are often too complex to be interpretable. 
Vice versa, a simpler pipeline with defined steps of operators may be easier to interpret but yield suboptimal prediction accuracy.
Finding the balance between pipeline complexity model interpretation and generalization remains a challenging task for AutoML application in biomedical big data.
With dataset selector, each pipeline individual of a TPOT generation during optimization holds lower complexity due to the selected subset's lower dimension compared to that of the entire dataset.
We hope that, with the complexity reduction from imposing a strongly-type GP template and dataset selector, a small loss in dataset-specific predictive accuracy can be compensated by considerable increase in interpretibility and generalizability.
In this study, the resulting TPOT-DS pipelines are more interpretable with only two simple optimized operators after the dataset selector: a transformer and a classifier.
In the case of the expression analysis, these pipelines also highlight two small sets of interconnected genes that contain candidates for MDD and related disorders.
Additionally, complexity reduction results in more efficient computation, which is strongly desirable in biomedical big data analysis.

A limitation of the DS analysis is the required pre-definition of subsets prior to executing TPOT-DS.
While this characteristic of an intelligent system is desirable when *a prior* knowledge on the biomedical data is available, it might pose as a challenge when this knowledge is inadequate, such as when analyzing data of a brand-new disease.
Nevertheless, one can perform a clustering method such as *k*-means to group features prior to performing TPOT-DS on the data. 
Another limitation of the current implementation of TPOT-DS is its restricted ability to select only one subset. 
A future design to support tree structures for Template will enable TPOT-DS to identify more than one subset that have high predictive power of the outcome.
A new operator that combines the data subsets will prove useful in this design.
Extensions of TPOT-DS will also involve overlapping subsets, which will require pipeline complexity reformulation beyond the total number of operators included in a pipeline.
Specifically, in the case of overlapping subsets, the number of features in the selected subset(s) is expected to be an element of the complexity calculation.
Extension of TPOT-DS to GWAS is straightforward.
However, because of the low predictive power of variants in current GWAS, alternative metrics beside accuracy, balanced accuracy or area under the receiving operator characteristic curve will need to be designed and included in the fitness function of TPOT's evolutionary algorithm.

[Conclusion]