<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Trang T. Le" />
  <meta name="author" content="Weixuan Fu" />
  <meta name="author" content="Jason H. Moore" />
  <meta name="dcterms.date" content="2018-11-08" />
  <meta name="keywords" content="tpot, automl, machine learning" />
  <title>Working Title</title>
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <link rel="stylesheet" href="github-pandoc.css" />
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!-- Insert Analytics Script Below -->
  <script>
  </script>
  <!-- End Analytics Script -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Working Title</h1>
</header>
<p><small><em> This manuscript (<a href="https://trang1618.github.io/tpot-ds-ms/v/8b471ec84df54a9409ff57f2a66287b36b66aa5f/">permalink</a>) was automatically generated from <a href="https://github.com/trang1618/tpot-ds-ms/tree/8b471ec84df54a9409ff57f2a66287b36b66aa5f">trang1618/tpot-ds-ms@8b471ec</a> on November 8, 2018. </em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Trang T. Le</strong><sup>☯</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0003-3737-6565">0000-0003-3737-6565</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/trang1618">trang1618</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/trang1618">trang1618</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 </small></p></li>
<li><p><strong>Weixuan Fu</strong><sup>☯</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0002-6434-5468">0000-0002-6434-5468</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/weixuanfu">weixuanfu</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/weixuanfu">weixuanfu</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 </small></p></li>
<li><p><strong>Jason H. Moore</strong><sup>†</sup><br> <img src="images/orcid.svg" alt="ORCID icon" width="13" height="13" /> <a href="https://orcid.org/0000-0002-5015-1099">0000-0002-5015-1099</a> · <img src="images/github.svg" alt="GitHub icon" width="13" height="13" /> <a href="https://github.com/EpistasisLab">EpistasisLab</a> · <img src="images/twitter.svg" alt="Twitter icon" width="13" height="13" /> <a href="https://twitter.com/moorejh">moorejh</a><br> <small> Department of Biostatistics, Epidemiology and Informatics, Institute for Biomedical Informatics, University of Pennsylvania, Philadelphia, PA 19104 · Funded by National Institute of Health Grant Nos. LM010098, LM012601 </small></p></li>
</ul>
<p><sup>☯</sup> — These authors contributed equally to this work.</p>
<p><sup>†</sup> — Direct correspondence to jhmoore@upenn.edu.</p>
<h2 id="abstract" class="page_break_before">Abstract</h2>
<h2 id="introduction">Introduction</h2>
<p>For many bioinformatics problems of classifying individuals into clinical categories from high-dimensional biological data, choosing a classifier is merely one step of the arduous process that leads to predictions. To detect patterns among features (<em>e.g.</em>, clinical variables) and their associations with the outcome (<em>e.g.</em>, clinical diagnosis), a data scientist typically has to design and test different complex machine learning frameworks that consist of data exploration, feature engineering, model selection and prediction. Automated machine learning (AutoML) systems were developed to automate this challenging and time-consuming process. These intelligent systems increase the accessibility and scalability of various machine learning applications by efficiently solving an optimization problem to discover pipelines that yield satisfactory outcomes, such as prediction accuracy. Consequently, AutoML allows data scientists to focus their effort in applying their expertise in other important research components such as developing meaningful hypotheses or communicating the results.</p>
<p>[other AutoML systems]</p>
<p>Tree-based Pipeline Optimization Tool (TPOT) is a genetic programming-based AutoML system that automates the laborious process of designing a machine learning pipeline to solve a supervised learning problem. At its core, TPOT uses genetic programming (GP) <span class="citation" data-cites="NopW1Vw3">[<a href="#ref-NopW1Vw3">1</a>]</span> to optimize a series of feature preprocessors and machine learning models with the objective of maximizing classification accuracy. While most AutoML systems primarily focus on model selection and hyperparameter optimization, TPOT also pays attention to feature selection and feature engineering in building a complete pipeline. Applying GP with the NSGA-II Pareto optimization <span class="citation" data-cites="iBP5Naag">[<a href="#ref-iBP5Naag">2</a>]</span>, TPOT optimizes the accuracy achieved by the pipeline while accounting for its complexity. Specifically, to automatically generate and optimize these machine learning pipelines, TPOT utilizes the Python package DEAP <span class="citation" data-cites="Gcs0HrMy">[<a href="#ref-Gcs0HrMy">3</a>]</span> to implement the GP algorithm.</p>
<p>Given no a priori knowledge about the problem, TPOT has been showed to frequently outperform standard machine learning analyses <span class="citation" data-cites="QkGSlAB3 JEn7WIoN">[<a href="#ref-QkGSlAB3">4</a>,<a href="#ref-JEn7WIoN">5</a>]</span>. Effort has been made to specialize TPOT for human genetics research, which results in a useful extended version of TPOT, TPOT-MDR, that features Multifactor Dimensionality Reduction and an Expert Knowledge Filter <span class="citation" data-cites="AvvI4W9K">[<a href="#ref-AvvI4W9K">6</a>]</span>. However, at the current stage, TPOT still requires great computational expense to analyze large datasets such as in genome-wide association studies or gene expression analyses. Consequently, application of TPOT on real-world datasets has been limited to small sets of features <span class="citation" data-cites="3LGbkjqK">[<a href="#ref-3LGbkjqK">7</a>]</span>.</p>
<p>In this work, we introduce two new features implemented in TPOT that helps increase the system’s scalability. First, the Dataset Selector (DS) allows the users to pass specific subsets of the features, reducing the computational expense of TPOT at the beginning of each pipeline to only evaluate on a smaller subset of data rather than the entire dataset. Consequently, DS makes TPOT applicable on large data sets by slicing the data into smaller sets of features (<em>e.g.</em> genes) and allowing genetic algorithm to select the best subset in the final pipeline. Second, Template enables the option for strongly typed GP, a method to enforce type constraints in genetic programming. By letting users specify a desired structure of the resulting machine learning pipeline, Template helps reduce TPOT computation time and potentially provide more interpretable results.</p>
<h2 id="methods">Methods</h2>
<h3 id="dataset-selector">Dataset Selector</h3>
<p>TPOT’s current operators include sets of feature pre-processors, feature transformers, feature selection techniques, and supervised classifiers and regressions. In this study, we introduce a new operator called Dataset Selector (DS) that enables biologically guided group-level feature selection. Specifically, taking place at the very first stage of the pipeline, DS passes only a specific subset of the features onwards. Hence, with DS, users can specify subsets of features of interest to reduce the feature space’s dimension at pipeline initialization. From predefined subsets of features, the DS operator allows TPOT to select the best subset that maximize average accuracy in k-fold cross validation (5-fold by default).</p>
<p>For example, in a gene expression analysis of major depressive disorder, a neuroscientist can specify collections of genes in pathways of interest and identify the important collection that helps predict the depression severity. Similarly, in a genome-wide association study of breast cancer, an analyst may assign variants in the data to different subsets of potentially related variants and detect the subset associated with the breast cancer diagnosis. In general, the DS operator allows for compartmentalization the feature space to smaller subsets based on <em>a priori</em> expert knowledge about the biomedical dataset. From here, TPOT selects the most relevant group of features, which can be utilized to motivate further analysis on that small group of features in biomedical research.</p>
<h3 id="template">Template</h3>
<p>Parallel with the establishment of the Dataset Selector operator, we now offer TPOT users the option to define a Template that provides a way to specify a desired structure for the resulting machine learning pipeline, which will reduce TPOT computation time and potentially provide more interpretable results.</p>
<p>Current implementation of Template supports linear pipelines, or path graphs, which are trees with two nodes (operators) of vertex degree 1, and the other <span class="math inline">\(n-2\)</span> nodes of vertex degree 2. Further, Template takes advantage of the strongly typed genetic programming framework that enforces data-type constraints <span class="citation" data-cites="KgFuJ0Jv">[<a href="#ref-KgFuJ0Jv">8</a>]</span> and imposes type-based restrictions on which element (<em>i.e.</em>, operator) type can be chosen at each node. In strongly typed genetic programming, while the fitness function and parameters remain the same, the initialization procedure and genetic operators (<em>e.g.</em>, mutation, crossover) must respect the enhanced legality constraints <span class="citation" data-cites="KgFuJ0Jv">[<a href="#ref-KgFuJ0Jv">8</a>]</span>. With a Template defined, each node in the tree pipeline is assigned one of the five major operator types: dataset selector, feature selection, feature transform, classifier or regressor. Moreover, besides the major operator types, each node can also be assigned more specifically as a method of an operator, such as decision trees for classifier. An example Template is Dataset selector → Feature transform → Decision trees.</p>
<h3 id="datasets">Datasets</h3>
<p>We apply TPOT with the new DS operator on both simulated datasets and a real world RNA-Seq gene expression dataset. With both real-world and simulated data, we hope to acquire a comprehensive view of the strengths and limitations of TPOT in the next generation sequencing domain.</p>
<h4 id="real-world-rna-seq-expression-data">Real-world RNA-Seq expression data</h4>
<p>We employed TPOT-DS on an RNA-Seq expression dataset of 78 individuals with major depressive disorder (MDD) and 79 healthy controls (HC) from Ref. <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">9</a>]</span>. Gene expression levels were quantified from reads of 19,968 annotated protein-coding genes and underwent a series of preprocessing steps including low read-count and outlier removal, technical and batch effect adjustment, and coefficient of variation filtering. Consequently, whole blood RNA-Seq measurements of 5,912 genes were obtained and are now used in the current study to test for association with MDD status. We use the 23 subsets of interconnected genes called depression gene modules (DGMs) identified from the RNA-Seq gene network module analysis <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">9</a>]</span> as input for the DS operator.</p>
<h4 id="simulation-methods">Simulation methods</h4>
<p>The simulated datasets were generated using the <code>R</code> package <code>privateEC</code>, which was designed to simulate realistic effects to be expected in gene expression or resting-state fMRI data. In the current study, to be consistent with the real expression dataset, we simulate interaction effect data with <em>m</em> = 200 individuals (100 cases and 100 controls) and <em>p</em>= 5,000 real-valued features with 4% functional (true positive association with outcome) for each training and testing set. Full details of the simulation approach can be found in Refs. <span class="citation" data-cites="p7dAO241 NKnMeQUs">[<a href="#ref-NKnMeQUs">10</a>,<a href="#ref-p7dAO241">9</a>]</span>. Briefly, the privateEC simulation induces a differential co-expression network random normal expression levels and permute the values of targeted features within the cases to generate interactions. Further, by imposing a large number of background features (no association with outcome), we seek to assess TPOT-DS’s performance in accommodating large numbers of non-predictive features.</p>
<p>To closely resemble the module size distribution in the RNA-Seq data, we first fit a <span class="math inline">\(\Gamma\)</span> distribution to the observed module sizes then sample from this distribution values for the simulated subset size, before the total number of features reaches 4,800 (number of background features). Then, the background features were randomly placed in each subset corresponding to its size. Also, for each subset <span class="math inline">\(S_i, i = 1, \dots, n\)</span>, a functional feature <span class="math inline">\(s_j\)</span> belongs to the subset with the probability <span id="eq:p_subset" style="display: inline-block; position: relative; width: 100%"><span class="math display">\[P(s_j \in S_i) \sim 1.618^{-i}\]</span><span style="position: absolute; right: 0em; top: 50%; line-height:0; text-align: right">(1)</span></span>  where 1.618 is an approximation of the golden ratio and yields a reasonable distribution of the functional features: they are more likely to be included in the earlier subsets (subset 1 and 2) than the later ones.</p>
<h3 id="performance-assessment">Performance assessment</h3>
<p>For each simulated and real-world dataset, after randomly splitting the entire data in two balanced smaller sets (75% training and 25% holdout), we trained TPOT-DS with the Template <code>Dataset Selector-Transformer-Classifier</code> on training data to predict class (e.g., diagnostic phenotype in real-world data) in the holdout set. We assess the performance of TPOT-DS by quantifying its ability to correctly select the most important subset (containing most functional features) in 100 replicates of TPOT runs on simulated data with known underlying truth. We also compare the out-of-sample accuracy of TPOT-DS’s exported pipeline on the holdout set with that of standard TPOT (with <code>Transformer-Classifier</code> Template, no DS operator) and XGBoost <span class="citation" data-cites="8w9fI63O">[<a href="#ref-8w9fI63O">11</a>]</span>, a fast and an efficient implementation of the gradient tree boosting method that has shown much utility in many winning Kaggle solutions <span class="citation" data-cites="1MHQyfXY">[<a href="#ref-1MHQyfXY">12</a>]</span> and been successfully incorporated in several neural network architectures <span class="citation" data-cites="19eUrsX1M 13as7dipI">[<a href="#ref-19eUrsX1M">13</a>,<a href="#ref-13as7dipI">14</a>]</span>. In the family of gradient boosted decision trees, XGBoost accounts for complex non-linear interaction structure among features and leverages gradient descents and boosting (sequential ensemble of weak classifiers) to effectively produce a strong prediction model. To obtain the optimal performance for this baseline model, we tune XGBoost hyperparameters with the <code>R</code> package <code>caret</code> <span class="citation" data-cites="6MvKCe21">[<a href="#ref-6MvKCe21">15</a>]</span> version 6.0-80 with the adaptive cross-validation algorithm.</p>
<h2 id="results">Results</h2>
<p>Our main goal is to test the performance of methods to identify features that discriminate between groups and optimize the classification accuracy.</p>
<h3 id="simulated-data">Simulated data</h3>
<p>We compare the accuracy of each method for <em>r</em> = 100 replicate simulated data sets with moderate interaction effect. These values of the effect size in the simulations generate adequately challenging data sets so that the methods’ accuracies stay moderate and do not cluster around 0.5 or 1. Each replicate data set is split into training and holdout. The TPOT-DS, standard TPOT and XGBoost models are built from the training dataset, and the trained model is then applied to the independent holdout data to obtain the generalization accuracy.</p>
<p>Our simulation design produces a reasonable distribution of the functional features, of which proportions in all subsets are shown in Table [S1]. According to Eq. <a href="#eq:p_subset">1</a>, the earlier the subset, the more functional features it has. Therefore, our first aim is to determine how well TPOT-DS can identify subset 1, which contains the largest number of informative features. With the specified template <code>Dataset Selector-Transformer-Classifier</code>, in 100 replications, TPOT-DS correctly selects subset 1 in the resulting pipeline [] times, with an average cross-validated accuracy on the training set of [] and out-of-sample accuracy of [] (Fig. [1]). Without DS, the standard TPOT and tuned XGBoost models respectively report a cross-validated accuracy of [] and [], and out-of-sample accuracy of [] and [0.565].</p>
<h3 id="rna-seq-expression-data">RNA-Seq expression data</h3>
<p>We apply standard TPOT, TPOT-DS and XGBoost to the RNA-Seq study of 78 major depressive disorder (MDD) subjects and 79 healthy controls described in <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">9</a>]</span>. The dataset contains 5,912 genes after preprocessing and filtering (see Methods for more detail). In 100 replications, TPOT-DS selects DGM-4 (282 genes) [] times to be the subset most predictive of the diagnosis status.</p>
<p><img src="images/real_43.svg" alt="Early results: TPOT-DS’s out-of-sample accuracy in RNA-Seq expression data with selected subset" />{height=“4in” “width=”5in&quot;}</p>
<h2 id="discussion">Discussion</h2>
<p>While the module scores of DGM-5 and DGM-17 were significantly associated with depressions severity measured by MADRS score <span class="citation" data-cites="p7dAO241">[<a href="#ref-p7dAO241">9</a>]</span>, we find with TPOT-DS that DGM-4 is largely predictive of the clinical diagnosis (MDD/HC) of each individual.</p>
<p>chalenging enough…</p>
<h3 id="limitations">Limitations</h3>
<p>[]</p>
<h3 id="future-works">Future works</h3>
<p>Extensions of TPOT-DS will involve overlapping subsets, which will require pipeline complexity reformulation beyond the total number of operators included in a pipeline. Also, a future design to support tree structures for Template will enable TPOT-DS to identify more than one subset that have high predictive power of the outcome.</p>
<h2 id="references" class="page_break_before">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs">
<div id="ref-NopW1Vw3">
<p>1. <strong>Genetic Programming: An Introduction: on the Automatic Evolution of Computer Programs and Its Applications</strong><br />
Wolfgang Banzhaf, Frank D. Francone, Robert E. Keller, Nordin Peter<br />
<em>ACM Digital Library</em> (1998) <a href="https://dl.acm.org/citation.cfm?id=280485">https://dl.acm.org/citation.cfm?id=280485</a></p>
</div>
<div id="ref-iBP5Naag">
<p>2. <strong>A fast and elitist multiobjective genetic algorithm: NSGA-II</strong><br />
K. Deb, A. Pratap, S. Agarwal, T. Meyarivan<br />
<em>IEEE Transactions on Evolutionary Computation</em> (2002-04) <a href="https://doi.org/bnw2vv">https://doi.org/bnw2vv</a><br />
DOI: <a href="https://doi.org/10.1109/4235.996017">10.1109/4235.996017</a></p>
</div>
<div id="ref-Gcs0HrMy">
<p>3. <strong>DEAP: Evolutionary Algorithms Made Easy</strong><br />
Félix-Antoine Fortin, François-Michel De Rainville, Marc-André Gardner, Marc Parizeau, Christian Gagné<br />
<em>Journal of Machine Learning Research</em> (2012) <a href="http://www.jmlr.org/papers/v13/fortin12a.html">http://www.jmlr.org/papers/v13/fortin12a.html</a></p>
</div>
<div id="ref-QkGSlAB3">
<p>4. <strong>Evaluation of a Tree-based Pipeline Optimization Tool for Automating Data Science</strong><br />
Randal S. Olson, Nathan Bartley, Ryan J. Urbanowicz, Jason H. Moore<br />
<em>Proceedings of the 2016 on Genetic and Evolutionary Computation Conference - GECCO ’16</em> (2016) <a href="https://doi.org/gfgqv2">https://doi.org/gfgqv2</a><br />
DOI: <a href="https://doi.org/10.1145/2908812.2908918">10.1145/2908812.2908918</a></p>
</div>
<div id="ref-JEn7WIoN">
<p>5. <strong>Identifying and Harnessing the Building Blocks of Machine Learning Pipelines for Sensible Initialization of a Data Science Automation Tool</strong><br />
Randal S. Olson, Jason H. Moore<br />
<em>arXiv</em> (2016-07-29) <a href="https://arxiv.org/abs/1607.08878v1">https://arxiv.org/abs/1607.08878v1</a></p>
</div>
<div id="ref-AvvI4W9K">
<p>6. <strong>Toward the automated analysis of complex diseases in genome-wide association studies using genetic programming</strong><br />
Andrew Sohn, Randal S. Olson, Jason H. Moore<br />
<em>Proceedings of the Genetic and Evolutionary Computation Conference on - GECCO ’17</em> (2017) <a href="https://doi.org/gfgqv3">https://doi.org/gfgqv3</a><br />
DOI: <a href="https://doi.org/10.1145/3071178.3071212">10.1145/3071178.3071212</a></p>
</div>
<div id="ref-3LGbkjqK">
<p>7. <strong>Integrated machine learning pipeline for aberrant biomarker enrichment (i-mAB): characterizing clusters of differentiation within a compendium of systemic lupus erythematosus patients</strong><br />
Trang T. Le, Nigel O. Blackwood, Jaclyn N. Taroni, Weixuan Fu, Matthew K. Breitenstein<br />
<em>arXiv</em> (2018-03-08) <a href="https://arxiv.org/abs/1803.04487v1">https://arxiv.org/abs/1803.04487v1</a></p>
</div>
<div id="ref-KgFuJ0Jv">
<p>8. <strong>Strongly Typed Genetic Programming</strong><br />
David J. Montana<br />
<em>Evolutionary Computation</em> (1995-06) <a href="https://doi.org/ct3mnb">https://doi.org/ct3mnb</a><br />
DOI: <a href="https://doi.org/10.1162/evco.1995.3.2.199">10.1162/evco.1995.3.2.199</a></p>
</div>
<div id="ref-p7dAO241">
<p>9. <strong>Identification and replication of RNA-Seq gene network modules associated with depression severity</strong><br />
Trang T. Le, Jonathan Savitz, Hideo Suzuki, Masaya Misaki, T. Kent Teague, Bill C. White, Julie H. Marino, Graham Wiley, Patrick M. Gaffney, Wayne C. Drevets, … Jerzy Bodurka<br />
<em>Translational Psychiatry</em> (2018-09-05) <a href="https://doi.org/gd7jx7">https://doi.org/gd7jx7</a><br />
DOI: <a href="https://doi.org/10.1038/s41398-018-0234-3">10.1038/s41398-018-0234-3</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/30185774">30185774</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC6125582">PMC6125582</a></p>
</div>
<div id="ref-NKnMeQUs">
<p>10. <strong>Differential co-expression network centrality and machine learning feature selection for identifying susceptibility hubs in networks with scale-free structure</strong><br />
Caleb A Lareau, Bill C White, Ann L Oberg, Brett A McKinney<br />
<em>BioData Mining</em> (2015-02-03) <a href="https://doi.org/gb5fpr">https://doi.org/gb5fpr</a><br />
DOI: <a href="https://doi.org/10.1186/s13040-015-0040-x">10.1186/s13040-015-0040-x</a> · PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/25685197">25685197</a> · PMCID: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4326454">PMC4326454</a></p>
</div>
<div id="ref-8w9fI63O">
<p>11. <strong>XGBoost</strong><br />
Tianqi Chen, Carlos Guestrin<br />
<em>Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining - KDD ’16</em> (2016) <a href="https://doi.org/gdp84q">https://doi.org/gdp84q</a><br />
DOI: <a href="https://doi.org/10.1145/2939672.2939785">10.1145/2939672.2939785</a></p>
</div>
<div id="ref-1MHQyfXY">
<p>12. <strong>Kaggle: Your Home for Data Science</strong><a href="https://www.kaggle.com/">https://www.kaggle.com/</a></p>
</div>
<div id="ref-19eUrsX1M">
<p>13. <strong>Short-Term Load Forecasting Using EMD-LSTM Neural Networks with a Xgboost Algorithm for Feature Importance Evaluation</strong><br />
Huiting Zheng, Jiabin Yuan, Long Chen<br />
<em>Energies</em> (2017-08-08) <a href="https://doi.org/gbtqbr">https://doi.org/gbtqbr</a><br />
DOI: <a href="https://doi.org/10.3390/en10081168">10.3390/en10081168</a></p>
</div>
<div id="ref-13as7dipI">
<p>14. <strong>A Novel Image Classification Method with CNN-XGBoost Model</strong><br />
Xudie Ren, Haonan Guo, Shenghong Li, Shilin Wang, Jianhua Li<br />
<em>Digital Forensics and Watermarking</em> (2017) <a href="https://doi.org/gfgvf3">https://doi.org/gfgvf3</a><br />
DOI: <a href="https://doi.org/10.1007/978-3-319-64185-0_28">10.1007/978-3-319-64185-0_28</a></p>
</div>
<div id="ref-6MvKCe21">
<p>15. <strong>Building Predictive Models inRUsing thecaretPackage</strong><br />
Max Kuhn<br />
<em>Journal of Statistical Software</em> (2008) <a href="https://doi.org/gdgzwf">https://doi.org/gdgzwf</a><br />
DOI: <a href="https://doi.org/10.18637/jss.v028.i05">10.18637/jss.v028.i05</a></p>
</div>
</div>
<script>
// AnchorJS minified version below.
// Source https://github.com/bryanbraun/anchorjs/blob/064abdd0987f305933ec4982af6d0c1cf2fd0814/anchor.js

/**
 * AnchorJS - v4.0.0 - 2017-06-02
 * https://github.com/bryanbraun/anchorjs
 * Copyright (c) 2017 Bryan Braun; Licensed MIT
 */
!function(A,e){"use strict";"function"==typeof define&&define.amd?define([],e):"object"==typeof module&&module.exports?module.exports=e():(A.AnchorJS=e(),A.anchors=new A.AnchorJS)}(this,function(){"use strict";function A(A){function e(A){A.icon=A.hasOwnProperty("icon")?A.icon:"",A.visible=A.hasOwnProperty("visible")?A.visible:"hover",A.placement=A.hasOwnProperty("placement")?A.placement:"right",A.class=A.hasOwnProperty("class")?A.class:"",A.truncate=A.hasOwnProperty("truncate")?Math.floor(A.truncate):64}function t(A){var e;if("string"==typeof A||A instanceof String)e=[].slice.call(document.querySelectorAll(A));else{if(!(Array.isArray(A)||A instanceof NodeList))throw new Error("The selector provided to AnchorJS was invalid.");e=[].slice.call(A)}return e}function n(){if(null===document.head.querySelector("style.anchorjs")){var A,e=document.createElement("style");e.className="anchorjs",e.appendChild(document.createTextNode("")),void 0===(A=document.head.querySelector('[rel="stylesheet"], style'))?document.head.appendChild(e):document.head.insertBefore(e,A),e.sheet.insertRule(" .anchorjs-link {   opacity: 0;   text-decoration: none;   -webkit-font-smoothing: antialiased;   -moz-osx-font-smoothing: grayscale; }",e.sheet.cssRules.length),e.sheet.insertRule(" *:hover > .anchorjs-link, .anchorjs-link:focus  {   opacity: 1; }",e.sheet.cssRules.length),e.sheet.insertRule(" [data-anchorjs-icon]::after {   content: attr(data-anchorjs-icon); }",e.sheet.cssRules.length),e.sheet.insertRule(' @font-face {   font-family: "anchorjs-icons";   src: url(data:n/a;base64,AAEAAAALAIAAAwAwT1MvMg8yG2cAAAE4AAAAYGNtYXDp3gC3AAABpAAAAExnYXNwAAAAEAAAA9wAAAAIZ2x5ZlQCcfwAAAH4AAABCGhlYWQHFvHyAAAAvAAAADZoaGVhBnACFwAAAPQAAAAkaG10eASAADEAAAGYAAAADGxvY2EACACEAAAB8AAAAAhtYXhwAAYAVwAAARgAAAAgbmFtZQGOH9cAAAMAAAAAunBvc3QAAwAAAAADvAAAACAAAQAAAAEAAHzE2p9fDzz1AAkEAAAAAADRecUWAAAAANQA6R8AAAAAAoACwAAAAAgAAgAAAAAAAAABAAADwP/AAAACgAAA/9MCrQABAAAAAAAAAAAAAAAAAAAAAwABAAAAAwBVAAIAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAMCQAGQAAUAAAKZAswAAACPApkCzAAAAesAMwEJAAAAAAAAAAAAAAAAAAAAARAAAAAAAAAAAAAAAAAAAAAAQAAg//0DwP/AAEADwABAAAAAAQAAAAAAAAAAAAAAIAAAAAAAAAIAAAACgAAxAAAAAwAAAAMAAAAcAAEAAwAAABwAAwABAAAAHAAEADAAAAAIAAgAAgAAACDpy//9//8AAAAg6cv//f///+EWNwADAAEAAAAAAAAAAAAAAAAACACEAAEAAAAAAAAAAAAAAAAxAAACAAQARAKAAsAAKwBUAAABIiYnJjQ3NzY2MzIWFxYUBwcGIicmNDc3NjQnJiYjIgYHBwYUFxYUBwYGIwciJicmNDc3NjIXFhQHBwYUFxYWMzI2Nzc2NCcmNDc2MhcWFAcHBgYjARQGDAUtLXoWOR8fORYtLTgKGwoKCjgaGg0gEhIgDXoaGgkJBQwHdR85Fi0tOAobCgoKOBoaDSASEiANehoaCQkKGwotLXoWOR8BMwUFLYEuehYXFxYugC44CQkKGwo4GkoaDQ0NDXoaShoKGwoFBe8XFi6ALjgJCQobCjgaShoNDQ0NehpKGgobCgoKLYEuehYXAAAADACWAAEAAAAAAAEACAAAAAEAAAAAAAIAAwAIAAEAAAAAAAMACAAAAAEAAAAAAAQACAAAAAEAAAAAAAUAAQALAAEAAAAAAAYACAAAAAMAAQQJAAEAEAAMAAMAAQQJAAIABgAcAAMAAQQJAAMAEAAMAAMAAQQJAAQAEAAMAAMAAQQJAAUAAgAiAAMAAQQJAAYAEAAMYW5jaG9yanM0MDBAAGEAbgBjAGgAbwByAGoAcwA0ADAAMABAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAH//wAP) format("truetype"); }',e.sheet.cssRules.length)}}this.options=A||{},this.elements=[],e(this.options),this.isTouchDevice=function(){return!!("ontouchstart"in window||window.DocumentTouch&&document instanceof DocumentTouch)},this.add=function(A){var i,o,s,c,r,a,h,l,u,d,f,g,p=[];if(e(this.options),"touch"===(g=this.options.visible)&&(g=this.isTouchDevice()?"always":"hover"),A||(A="h2, h3, h4, h5, h6"),0===(i=t(A)).length)return this;for(n(),o=document.querySelectorAll("[id]"),s=[].map.call(o,function(A){return A.id}),r=0;r<i.length;r++)if(this.hasAnchorJSLink(i[r]))p.push(r);else{if(i[r].hasAttribute("id"))c=i[r].getAttribute("id");else if(i[r].hasAttribute("data-anchor-id"))c=i[r].getAttribute("data-anchor-id");else{u=l=this.urlify(i[r].textContent),h=0;do{void 0!==a&&(u=l+"-"+h),a=s.indexOf(u),h+=1}while(-1!==a);a=void 0,s.push(u),i[r].setAttribute("id",u),c=u}d=c.replace(/-/g," "),(f=document.createElement("a")).className="anchorjs-link "+this.options.class,f.href="#"+c,f.setAttribute("aria-label","Anchor link for: "+d),f.setAttribute("data-anchorjs-icon",this.options.icon),"always"===g&&(f.style.opacity="1"),""===this.options.icon&&(f.style.font="1em/1 anchorjs-icons","left"===this.options.placement&&(f.style.lineHeight="inherit")),"left"===this.options.placement?(f.style.position="absolute",f.style.marginLeft="-1em",f.style.paddingRight="0.5em",i[r].insertBefore(f,i[r].firstChild)):(f.style.paddingLeft="0.375em",i[r].appendChild(f))}for(r=0;r<p.length;r++)i.splice(p[r]-r,1);return this.elements=this.elements.concat(i),this},this.remove=function(A){for(var e,n,i=t(A),o=0;o<i.length;o++)(n=i[o].querySelector(".anchorjs-link"))&&(-1!==(e=this.elements.indexOf(i[o]))&&this.elements.splice(e,1),i[o].removeChild(n));return this},this.removeAll=function(){this.remove(this.elements)},this.urlify=function(A){var t=/[& +$,:;=?@"#{}|^~[`%!'<>\]\.\/\(\)\*\\]/g;return this.options.truncate||e(this.options),A.trim().replace(/\'/gi,"").replace(t,"-").replace(/-{2,}/g,"-").substring(0,this.options.truncate).replace(/^-+|-+$/gm,"").toLowerCase()},this.hasAnchorJSLink=function(A){var e=A.firstChild&&(" "+A.firstChild.className+" ").indexOf(" anchorjs-link ")>-1,t=A.lastChild&&(" "+A.lastChild.className+" ").indexOf(" anchorjs-link ")>-1;return e||t||!1}}return A});

// Enable links for selected headers
var anchors = new AnchorJS();
anchors.add("h2, h3, h4")
</script>
<!-- Enable Hypothesis annotation. https://web.hypothes.is/ -->
<script src="https://hypothes.is/embed.js" async></script>
</body>
</html>
